src/v/storage/compaction_reducers.cc

```diff
@@ -287,6 +287,19 @@ ss::future<ss::stop_iteration> copy_data_segment_reducer::do_compaction(
     if (to_copy == std::nullopt) {
         co_return stop_t::no;
     }
+    if (_compacted_idx && is_compactible(to_copy.value())) {
```

Comment:
This method builds a new segment, and the keys in the compacted index should reflect that, no? It's possible that the new segment has virtually no keys because all of the new values were written in newer segments, in which case keeping the existing compacted index would be wasteful.

To your point, we could just delete the compacted index and instead rely on a subsequent round of self compaction to create the indexes. OTOH I'm thinking that windowed compaction would be a sort of supserset of self compaction: if you've done windowed compaction, there's no need to self compact because the indexes are already there.